WIP.

vanzin opened the pull request:
Test with this, for example, no need for a cluster or a DFS at all:
./bin/spark-shell --master 'local-cluster[4,1,1024]' \
  --conf spark.shuffle.io.plugin.class=org.apache.spark.shuffle.dfs.DFSShuffleDataIO \
  --conf spark.shuffle.dfs.basePath=file:/tmp/dfs.io

Commit:
WIP.

Test with this, for example, no need for a cluster or a DFS at all:
./bin/spark-shell --master 'local-cluster[4,1,1024]' \
  --conf spark.shuffle.io.plugin.class=org.apache.spark.shuffle.dfs.DFSShuffleDataIO \
  --conf spark.shuffle.dfs.basePath=file:/tmp/dfs.io

yifeih committed:
Update ShuffleLocation to be part of the read API too

ifilonenko commented on core/src/main/scala/org/apache/spark/shuffle/ShufflePartitionWriterOutputStream.scala:
```diff
@@ -20,38 +20,36 @@ package org.apache.spark.shuffle
+    var bufferInputStream: InputStream = new ByteBufferInputStream(buffer)
     if (underlyingOutputStream == null) {
-      underlyingOutputStream = partitionWriter.openPartitionStream()
+      underlyingOutputStream = serializerManager.wrapStream(blockId,
```
Good catch here

mccheah replied:
I think doing it this way makes it so that Spark is opinionated about writing the stream with IO encryption and compression. Should it not be the responsibility for the underlying writer to decide its own encryption and compression?

mccheah replied:
Ah never mind because this has to match the read side everywhere.

mccheah replied:
What makes this interesting is that if the underlying implementation decides to do its own encryption and compression, we're effectively wasting work here that the underlying implementation is going to do.

vanzin replied:
Users can always just not enable shuffle encryption; compression is enabled by default, but can also be disabled.

ifilonenko commented:
@vanzin any chance you could make a PR against `ifilonenko:SPARK-25299-v3` for the fixes outside of the DFS implementation? else, I'd be happy to cherry-pick those changes as they seem quite critical

vanzin committed:
Changes to ByteBuffer and serialization logic

- add exceptions to API
- fix write side: use serializer manager, fix usage of ByteBuffer.
- misc compilation / style fixes to get things to build.

ifilonenko committed:
resolve some comments regarding BlockManager and slight style

yifeih committed:
Fix UnsafeShuffleWriter (#15)

* compiles
`

* fix UnsafeShuffleWriter

* remove unnecessary changes

Commit:
Add cleanup callback to shuffle IO interface.
